### PR TITLE
Implement Num from num-traits

### PR DESCRIPTION
**Pull request by SamWilsn**
Implementing the [`Num`](https://docs.rs/num-traits/*/num_traits/trait.Num.html) trait makes the `U256` (and other) number types easier to work with, specifically with `from_str_radix`.

### CODE REVIEW

File: primitive-types/impls/num-traits/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+		impl $crate::num_traits::Num for $name {
+			type FromStrRadixErr = $crate::FromStrRadixErr;
+
+			fn from_str_radix(txt: &str, radix: u32) -> Result<Self, Self::FromStrRadixErr> {
+				let parsed = match radix {
+					10 => Self::from_dec_str(txt)?,
+					16 => core::str::FromStr::from_str(txt)?,
+					_ => return Err(Self::FromStrRadixErr::unsupported()),
+				};
```
Comment (ordian) [inline primitive-types/impls/num-traits/src/lib.rs:121]: do you think it'd be good to add this as a static method to `uint` to mimic std types?
https://doc.rust-lang.org/std/?search=from_str_radix

File: primitive-types/impls/num-traits/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		match self {
+			FromStrRadixErrSrc::Dec(d) => write!(f, "{}", d),
+			FromStrRadixErrSrc::Hex(h) => write!(f, "{}", h),
+		}
+	}
+}
+
+#[derive(Debug)]
+pub struct FromStrRadixErr {
```
Comment (ordian) [inline primitive-types/impls/num-traits/src/lib.rs:40]: thank you for this clean design with opaque error type,
our policy is to have all new public types/methods/API documented (`///`)

**Review by ordian — COMMENTED** [review]
Thanks!

I usually don't like adding more features based on third-party libs, because it their breaking changes will be our breaking changes (semver). However I don't really mind in this case if you have a good use-case for it.

File: primitive-types/impls/num-traits/src/lib.rs (hunk 1 above)
Comment (SamWilsn) [reply primitive-types/impls/num-traits/src/lib.rs:121]: Yeah, I can do that! I'll move the error type into `uint`?

**Comment by SamWilsn** [comment]
I think, since this is gated behind features, you could do a `num-traits-03` feature when version `0.3` is released.

I'll get the changes done tomorrow for the other comments.

**Review by ordian — APPROVED** [review]
Looks good, thanks!

(a note for maintainers: this is a non-breaking change)

File: primitive-types/impls/num-traits/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! num-traits support for uint.
+
+#![cfg_attr(not(feature = "std"), no_std)]
```
Comment (niklasad1) [inline primitive-types/impls/num-traits/src/lib.rs:11]: I don't understand why is this needed?

**Review by niklasad1 — APPROVED** [review]
LGTM

File: primitive-types/impls/num-traits/src/lib.rs (hunk 3 above)
Comment (SamWilsn) [reply primitive-types/impls/num-traits/src/lib.rs:11]: Bit of a leftover from moving `from_str_radix` into the `uint` crate. I don't believe this crate needs it ever, but there is a feature that enables `std` in `uint` and `num-traits`.

File: primitive-types/src/lib.rs [hunk 4]
```diff
@@ -22,6 +22,7 @@ use fixed_hash::{construct_fixed_hash, impl_fixed_hash_conversions};
 #[cfg(feature = "scale-info")]
 use scale_info::TypeInfo;
 use uint::{construct_uint, uint_full_mul_reg};
+pub use uint::{FromStrRadixErr, FromStrRadixErrKind};
```
Comment (ordian) [inline primitive-types/src/lib.rs:25]: hmm, I overlooked this line, seems redundant
we could return `uint::FromStrRadixErr` in `impls/num-traits` instead of `$crate::...`

Comment (SamWilsn) [reply primitive-types/src/lib.rs:25]: I'm not super familiar with the macro hygiene rules, but what if the downstream crate did a `extern crate uint as floop`?

Comment (ordian) [reply primitive-types/src/lib.rs:25]: yeah, fair point, we can do what `impl/rlp` does, but still this like seems redundant

**Comment by ordian** [comment]
I'll address the nits in a follow-up.
